File: python/src/resp_bench/metrics/collector.py
"""Latency metrics collection.

Single-event-loop design: no locks are needed because ``record`` runs to
completion without awaiting, so concurrent worker coroutines never interleave
inside it. Latencies are clamped to 600s before recording and errors are
counted but not recorded into the histogram -- matching the other engines.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Dict, Optional

from ..command.command import CommandResult
from .hdr_encoder import HIGHEST_TRACKABLE_VALUE, new_histogram


class CommandMetrics:
    def __init__(self, command_name: str) -> None:
        self.command_name = command_name
        self.requests = 0
        self.errors = 0
        # Created eagerly (like the Java reference) so the NDJSON hdr block and
        # summary are always present, even for a command that only ever errors
        # (an empty histogram reports count 0 and zero percentiles).
        self._histogram = new_histogram()

    def record(self, result: CommandResult) -> None:
        self.requests += 1
        if result.success:
            latency = min(result.latency_micros, HIGHEST_TRACKABLE_VALUE)
            self._histogram.record_value(latency)
        else:
            self.errors += 1

    @property
    def histogram(self):
        return self._histogram

    def count(self) -> int:
        return self._histogram.get_total_count()

    def min(self) -> int:
        return self._histogram.get_min_value()

    def max(self) -> int:
        return self._histogram.get_max_value()

    def percentile(self, pct: float) -> int:
        return self._histogram.get_value_at_percentile(pct)


class MetricsCollector:
    def __init__(self) -> None:
        self.command_metrics: Dict[str, CommandMetrics] = {}
        self.total_requests = 0
        self.total_errors = 0
        self.start_time: Optional[datetime] = None
        self.end_time: Optional[datetime] = None

    def start(self) -> None:
        self.start_time = datetime.now(timezone.utc)

    def stop(self) -> None:
        self.end_time = datetime.now(timezone.utc)

    def record(self, result: CommandResult) -> None:
        self.total_requests += 1
        if not result.success:
            self.total_errors += 1

        metrics = self.command_metrics.get(result.command_name)
        if metrics is None:
            metrics = CommandMetrics(result.command_name)
            self.command_metrics[result.command_name] = metrics
        metrics.record(result)

    def duration_millis(self) -> int:
        if not self.start_time or not self.end_time:
            return 0
        return int((self.end_time - self.start_time).total_seconds() * 1000)

File: node/src/metrics/collector.ts
/**
 * Latency metrics collection.
 *
 * Single-event-loop design: no locks are needed because `record` runs to
 * completion without awaiting, so concurrent workers never interleave inside it.
 * Latencies are clamped to 600s before recording, and errors are counted but not
 * recorded into the histogram -- matching the other engines.
 */

import type { CommandResult } from '../command/command.js';
import { HIGHEST_TRACKABLE_VALUE, newHistogram, type Histogram } from './hdrHistogram.js';

export class CommandMetrics {
  readonly commandName: string;
  requests = 0;
  errors = 0;
  /**
   * Created eagerly (like the Java reference) so the NDJSON `hdr` block and
   * summary are always present, even for a command that only ever errors -- an
   * empty histogram reports count 0 and zero percentiles.
   */
  readonly histogram: Histogram = newHistogram();

  constructor(commandName: string) {
    this.commandName = commandName;
  }

  record(result: CommandResult): void {
    this.requests += 1;
    if (result.success) {
      this.histogram.recordValue(Math.min(result.latencyMicros, HIGHEST_TRACKABLE_VALUE));
    } else {
      this.errors += 1;
    }
  }

  count(): number {
    return this.histogram.totalCount;
  }

  /**
   * Lowest recorded latency, matching Java's `Histogram.getMinValue()`.
   *
   * Deliberately NOT `minNonZeroValue`: that skips a legitimately recorded 0us
   * sample, and on an empty histogram it returns Number.MAX_SAFE_INTEGER, which
   * would land in the NDJSON as a nonsense min. `getValueAtPercentile(0)` returns
   * 0 when empty and Java's `getMinValue()` otherwise -- verified equal for
   * 1us..599s (and it is what the Ruby encoder uses).
   */
  min(): number {
    return this.histogram.getValueAtPercentile(0);
  }

  /**
   * Highest recorded latency, matching Java's `Histogram.getMaxValue()`.
   *
   * Deliberately NOT `maxValue`: hdr-histogram-js returns the raw recorded
   * sample there, while Java returns the *bucket's* highest equivalent value. At
   * 3 significant figures those diverge above ~1000us -- recording 50000us gives
   * 50000 in JS but 50015 in Java -- which would make summary.max quietly
   * incomparable across engines. `getValueAtPercentile(100)` is Java's value.
   */
  max(): number {
    return this.histogram.getValueAtPercentile(100);
  }

  percentile(pct: number): number {
    return this.histogram.getValueAtPercentile(pct);
  }
}

export class MetricsCollector {
  readonly commandMetrics = new Map<string, CommandMetrics>();
  totalRequests = 0;
  totalErrors = 0;
  private startTimeMs: number | null = null;
  private endTimeMs: number | null = null;

  start(): void {
    this.startTimeMs = Date.now();
  }

  stop(): void {
    this.endTimeMs = Date.now();
  }

  record(result: CommandResult): void {
    this.totalRequests += 1;
    if (!result.success) this.totalErrors += 1;

    let metrics = this.commandMetrics.get(result.commandName);
    if (metrics === undefined) {
      metrics = new CommandMetrics(result.commandName);
      this.commandMetrics.set(result.commandName, metrics);
    }
    metrics.record(result);
  }

  startTime(): number {
    return this.startTimeMs ?? 0;
  }

  endTime(): number {
    return this.endTimeMs ?? 0;
  }

  durationMillis(): number {
    if (this.startTimeMs === null || this.endTimeMs === null) return 0;
    return this.endTimeMs - this.startTimeMs;
  }
}
